Fixed #29177: FKs accessible in subsequent migrations for unmanaged models

michalnik opened the pull request:
#### Work done
1. I added regression tests for the scenario described in ticket #29177 in the test_autodetector_unmanaged module.
2. I implemented a fix in django.db.migrations.autodetector.
3. Finally, I added also unittests for Autodetector covering problems with unmanaged models and foreign keys

#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-29177

#### Branch description
Provide a concise overview of the issue or rationale behind the proposed changes.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.

michalnik commented:
@brosner Hello, sorry to bother you, but I'm not sure how to proceed with this PR. From my side, it's finished, but I'm not sure if I need to add unittests. On the Discord server, I was advised to write regression tests, and those are done. I see some failed checks here, but I have no idea what they're about.

coolbootscoder commented:
@michalnik I'm not a maintainer but I think you will need to resolve the conflicts before a reviewer looks at the pull request. After a successful review and the ticket is marked for checkin, then a merger will look at the pull request.

michalnik committed:
Added tests for #29177.

Integration tests part (test_autodetector_unmanaged)
1) On creation of a new managed=False model in migration with FK.
2) On addition of a FK field to already created managed=False model.
3) On removing of a FK field from already created managed=False model.

Unit tests part (test_autodetector)
1) add FK to unmanged model
2) remove FK from unmanaged model
3) rename FK within unmanaged model

michalnik commented:
Sorry, currently I am struggling with anything from my previous commits did not work as expected.

michalnik committed:
Fixed missing fields while unmanaged model is used in migration.

michalnik committed:
Fixed black linter after rebase from upstream.

jacobtylerwalls commented:
Thanks for the PR (and sorry for the delayed review)! This issue was fixed in the related #18651.

michalnik commented:
> Thanks for the PR (and sorry for the delayed review)! This issue was fixed in the related #18651.

Hi! Thanks for taking the time to look at this.

I understand why this PR was closed as a duplicate of the other fix.

Since I've been working on this issue for quite a while, I'd still really appreciate any feedback on my implementation or the overall approach whenever someone has a chance. Even if this PR won't be merged, I'd like to know whether my reasoning, tests, and solution are considered sound, so I can improve future contributions.

Thanks!

jacobtylerwalls commented:
Sure. The approach was sound. I would have just added new tests only for test_autodetector.py in the spirit of similar changes that didn't add new integration test files
